src/bun.js/net.exports.js

```diff
@@ -53,9 +53,246 @@ export function isIP(s) {
+        family,
+        hints,
+        lookup,
+        noDelay,
+        keepAlive,
+        keepAliveInitialDelay,
+      } = port;
+    }
+    const self = this;
+    let ended = false;
+    self.connecting = true;
+    self.remotePort = port;
+    if (connectListener) self.on("connect", connectListener);
+    Bun.connect({
+      hostname: host || "localhost",
+      port: port,
+      socket: {
+        close,
+        data(socket, buffer) {
+          self.bytesRead += buffer.length;
+          if (self.#readQueue.length > 0 || !self.push(buffer)) {
+            self.#readQueue.push(buffer);
+          }
+        },
+        drain,
+        end: close,
+        error(socket, error) {
+          const { callback } = self.#writeQueue || {};
+          if (callback) {
+            self.#writeQueue = null;
+            callback(error);
+          }
+          self.emit("error", error);
+        },
+        open(socket) {
+          socket.timeout(self.timeout);
+          self.#socket = socket;
+          self.connecting = false;
+          self.emit("connect");
+          drain(socket);
+        },
+        timeout() {
+          self.emit("timeout");
+        },
+      },
+    });
+    return this;
+
+    function close() {
+      if (ended) return;
+      ended = true;
+      if (self.#readQueue.length > 0) {
+        self.#readQueue.push(null);
+      } else {
+        self.push(null);
+      }
+    }
+
+    function drain(socket) {
+      const { chunk, callback } = self.#writeQueue || {};
+      if (callback) {
+        const written = socket.write(chunk);
+        self.bytesWritten += written;
+        if (written < chunk.length) {
+          self.#writeQueue.chunk = chunk.slice(written);
+        } else {
+          self.#writeQueue = null;
+          callback(null);
+        }
+      }
+    }
+  }
+
+  _destroy(err, callback) {
+    this.#socket?.shutdown();
+    callback(err);
+  }
+
+  _final(callback) {
+    this.#socket.end();
+    callback();
+  }
+
+  get localAddress() {
+    return "127.0.0.1";
+  }
+
+  get localFamily() {
+    return "IPv4";
+  }
+
+  get localPort() {
+    return this.#socket?.localPort;
+  }
+
+  get pending() {
+    return this.connecting;
+  }
+
+  _read(size) {
+    const buffers = this.#readQueue;
+    const len = buffers.length;
+    if (len == 0) return;
+    for (var i = 0; i < len; i++) {
+      if (!this.push(buffers[i])) break;
+    }
+    if (i == len) {
+      buffers.length = 0;
+    } else {
+      this.#readQueue = buffers.slice(i);
+    }
+  }
+
+  get readyState() {
+    if (this.connecting) return "opening";
+    if (this.readable) {
+      return this.writable ? "open" : "readOnly";
+    } else {
+      return this.writable ? "writeOnly" : "closed";
+    }
+  }
+
+  ref() {
+    this.#socket?.ref();
+  }
+
+  get remoteAddress() {
+    return this.#socket.remoteAddress;
+  }
+
+  get remoteFamily() {
+    return "IPv4";
+  }
+
+  resetAndDestroy() {
+    this.#socket?.shutdown();
```

Comment:
`shutdown` doesn't destroy the socket

It prevents reads/writes from happening without actually terminating the connection

Reply:
changed to `.end()` - https://github.com/oven-sh/bun/pull/1701/commits/81b7104e6517d030ce8aea87609f4b938509a1c3